src/dir.rs

```diff
@@ -0,0 +1,198 @@
+//
+// `Dir` is safe to pass from one thread to another, as it's not reference-counted.
+unsafe impl Send for Dir {}
+
+impl<'d> IntoIterator for &'d mut Dir {
+    type Item = Result<Entry>;
+    type IntoIter = Iter<'d>;
+    fn into_iter(self) -> Self::IntoIter {
+        Iter(self)
+    }
+}
+
+impl AsRawFd for Dir {
+    fn as_raw_fd(&self) -> RawFd {
+        unsafe { libc::dirfd(self.0) }
+    }
+}
+
+impl fmt::Debug for Dir {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Dir")
+            .field("fd", &self.as_raw_fd())
+            .finish()
+    }
+}
+
+impl Drop for Dir {
+    fn drop(&mut self) {
+        unsafe { libc::closedir(self.0) };
+    }
+}
+
+#[derive(Debug)]
+pub struct Iter<'d>(&'d mut Dir);
+
+impl<'d> Iterator for Iter<'d> {
+    type Item = Result<Entry>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            // Note: POSIX specifies that portable applications should dynamically allocate a
+            // buffer with room for a `d_name` field of size `pathconf(..., _PC_NAME_MAX)` plus 1
+            // for the NUL byte. It doesn't look like the std library does this; it just uses
+            // fixed-sized buffers (and libc's dirent seems to be sized so this is appropriate).
+            // Probably fine here too then.
+            let mut ent: Entry = Entry(::std::mem::uninitialized());
+            let mut result = ptr::null_mut();
+            if let Err(e) = Errno::result(readdir_r((self.0).0, &mut ent.0, &mut result)) {
+                return Some(Err(e));
+            }
+            if result == ptr::null_mut() {
+                return None;
+            }
+            assert_eq!(result, &mut ent.0 as *mut dirent);
+            return Some(Ok(ent));
+        }
+    }
+}
+
+impl<'d> Drop for Iter<'d> {
+    fn drop(&mut self) {
+        unsafe { libc::rewinddir((self.0).0) }
+    }
+}
+
+/// A directory entry, similar to `std::fs::DirEntry`.
+/// Note that unlike the std version, this may represent the `.` or `..` entries.
+#[derive(Copy, Clone)]
+pub struct Entry(dirent);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Type {
+    Fifo,
+    CharacterDevice,
+    Directory,
+    BlockDevice,
+    File,
+    Symlink,
+    Socket,
+}
+
+impl Entry {
+    /// Returns the inode number (`d_ino`) of the underlying `dirent`.
+    #[cfg(any(target_os = "macos",
+              target_os = "ios",
+              target_os = "linux",
+              target_os = "emscripten",
+              target_os = "android",
+              target_os = "solaris",
+              target_os = "haiku",
+              target_os = "l4re",
+              target_os = "fuchsia"))]
+    pub fn ino(&self) -> u64 {
+        self.0.d_ino as u64
+    }
+
+    /// Returns the inode number (`d_fileno`) of the underlying `dirent`.
+    #[cfg(any(target_os = "freebsd",
+              target_os = "openbsd",
+              target_os = "bitrig",
```

Comment:
bitrig is dead.  You can remove it.  Also, the others should be alphabetically ordered.  Does Nix support any OSes that can't implement `ino`?  If not, you can simplify this with `cfg_if`, like this:
```rust
pub fn ino(&self) -> u64 {
  cfg_if! {
    if #[cfg(any(target_os = ...))] {
       self.0.d_ino as u64
    } else {
      self.0.d_filenno as u64
    }
  }
}

Reply:
Hmm, that's neat, but I'm having trouble getting it to work. Tried a couple variations. Any ideas?

```rust
    pub fn ino(&self) -> u64 {
        cfg_if! {
            if #[cfg(any(target_os = "android",
                         target_os = "emscripten",
                         target_os = "fuchsia",
                         target_os = "haiku",
                         target_os = "ios",
                         target_os = "l4re",
                         target_os = "linux",
                         target_os = "macos",
                         target_os = "solaris"))] {
                self.0.d_ino as u64
            } else {
                self.0.d_fileno as u64
            }
        }
    }
```

produces

```
error: expected one of `!` or `::`, found `.`
   --> src/dir.rs:158:21
    |
158 |                 self.0.d_ino as u64
    |                     ^ expected one of `!` or `::` here
```

and

```rust
    cfg_if! {
        if #[cfg(any(target_os = "android",
                     target_os = "emscripten",
                     target_os = "fuchsia",
                     target_os = "haiku",
                     target_os = "ios",
                     target_os = "l4re",
                     target_os = "linux",
                     target_os = "macos",
                     target_os = "solaris"))] {
            pub fn ino(&self) -> u64 { self.0.d_ino as u64 }
        } else {
            pub fn ino(&self) -> u64 { self.0.d_fileno as u64 }
        }
    }
```

produces a bunch of errors, starting with:

```
error: expected one of `::` or `:`, found `)`
   --> src/dir.rs:175:29
    |
175 |             pub fn ino(&self) -> u64 { self.0.d_ino as u64 }
    |                             ^ expected one of `::` or `:` here
```

Reply:
Looks like `cfg_if` doesn't work with expressions, only items.  An item is something like a module or a function.
https://docs.rs/cfg-if/0.1.4/cfg_if/macro.cfg_if.html
https://doc.rust-lang.org/reference/items.html

Reply:
Hmm, function didn't work either (second try above). I got it to work by having a separate `impl` for this vs the other stuff, but then the two impls are separate in the documentation, which is annoying.

I think I'm giving up on `cfg_if!` here, but I can do `#[cfg(any(target_os = "bar", target_os="foo"))]` vs `#[cfg(not(any(target_os = "bar", target_os="foo")))]` so it's obviously comprehensive.